File: runme/mcgen/local/datan/solsrc/s2sd.f
      PROGRAM S2SD
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION XPL(1000),YPL(1000)
      CHARACTER*75 TX,TY,CAPT
      PARAMETER(NPL=1000)
C identify program to user
      WRITE(*,*)' Program S2D provides interactively graphs of'
      WRITE(*,*)' statistical functions of a continuous variable'
      WRITE(*,*)' '
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - standardized normal distribution'
      WRITE(*,*)' 2 - normal distribution'
      WRITE(*,*)' 3 - chi-squared distribution'
      WRITE(*,*)' 4 - F distribution'
      WRITE(*,*)' 5 - t distribution'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - probability density'
      WRITE(*,*)' 2 - distribution function'
      WRITE(*,*)' >'
      READ(*,*)ISWIT1
      IF(ISWIT.EQ.1) THEN
C standardized normal
        XMIN=-3.D0
        XMAX=3.D0
        YMIN=0.D0
        IF(ISWIT1.EQ.1) THEN
          YMAX=0.5D0
        ELSE
          YMAX=1.D0
        END IF
        TX='x'
        LTX=1
        IF(ISWIT1.EQ.1) THEN
          TY='&f@(x)'
        ELSE IF(ISWIT1.EQ.2) THEN
          TY='&F@(x)'
        END IF
        LTY=6
        CAPT='standardized normal distribution'
        LCA=32
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 10 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          IF(ISWIT1.EQ.1) THEN
            YPL(I)=SDSTNR(XPL(I))
          ELSE IF(ISWIT1.EQ.2) THEN
            YPL(I)=SCSTNR(XPL(I))
          END IF
10      CONTINUE
      ELSE IF(ISWIT.EQ.2) THEN
C normal
        WRITE(*,*)' Enter x0'
        WRITE(*,*)' >'
        READ(*,*)X0
        WRITE(*,*)' Enter sigma (> 0.)'
        WRITE(*,*)' >'
        READ(*,*)SIGMA
        XMIN=-3.D0*SIGMA+X0
        XMAX=3.D0*SIGMA+X0
        YMIN=0.D0
        IF(ISWIT1.EQ.1) THEN
          YMAX=1.2D0*SDNORM(X0,X0,SIGMA)
        ELSE
          YMAX=1.D0
        END IF
        TX='x'
        LTX=1
        IF(ISWIT1.EQ.1) THEN
          TY='f(x)'
        ELSE IF(ISWIT1.EQ.2) THEN
          TY='F(x)'
        END IF
        LTY=4
        CAPT='normal distribution with x_0# = ******, &s@ = ******'
        WRITE(CAPT(33:38),'(F6.2)')X0
        WRITE(CAPT(47:52),'(F6.2)')SIGMA
        LCA=52
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 20 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          IF(ISWIT1.EQ.1) THEN
            YPL(I)=SDNORM(XPL(I),X0,SIGMA)
          ELSE IF(ISWIT1.EQ.2) THEN
            YPL(I)=SCNORM(XPL(I),X0,SIGMA)
          END IF
20      CONTINUE
      ELSE IF(ISWIT.EQ.3) THEN
C chi-squared
        WRITE(*,*)' Enter n (>= 1)'
        WRITE(*,*)' >'
        READ(*,*)N
        XMIN=0.D0
        XMAX=SQCHI2(.999D0,N)
        IF(N.EQ.1) XMIN=.00001D0
        YMIN=0.D0
        IF(ISWIT1.EQ.1) THEN
          IF(N.EQ.1) YMAX=2.D0
          IF(N.EQ.2) YMAX=.5D0
          IF(N.GT.2) YMAX=1.2D0*SDCHI2(DBLE(N-1),N)
        ELSE IF(ISWIT1.EQ.2) THEN
          YMAX=1.D0
        END IF
        TX='&c^2'
        LTX=4
        IF(ISWIT1.EQ.1) THEN
          TY='f(&c@^2#)'
        ELSE IF(ISWIT1.EQ.2) THEN
          TY='F(&c@^2#)'
        END IF
        LTY=9
        CAPT='&c@^2# distribution with n = ***'
        WRITE(CAPT(30:32),'(I3)')N
        LCA=32
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 30 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          IF(ISWIT1.EQ.1) THEN
            YPL(I)=SDCHI2(XPL(I),N)
          ELSE IF(ISWIT1.EQ.2) THEN
            YPL(I)=SCCHI2(XPL(I),N)
          END IF
30      CONTINUE
      ELSE IF(ISWIT.EQ.4) THEN
C F distribution
        WRITE(*,*)' Enter f1 (>= 1)'
        WRITE(*,*)' >'
        READ(*,*)NF1
        WRITE(*,*)' Enter f2 (>= 1)'
        WRITE(*,*)' >'
        READ(*,*)NF2
        XMIN=-0.D0
        XMAX=3.D0
        YMIN=0.D0
        YMAX=1.D0
        TX='F'
        LTX=1
        IF(ISWIT1.EQ.1) THEN
          TY='f(F)'
        ELSE IF(ISWIT1.EQ.2) THEN
          TY='F(F)'
        END IF
        LTY=4
        CAPT='F distribution with f_1# = ***, f_2# = ***'
        WRITE(CAPT(28:30),'(I3)')NF1
        WRITE(CAPT(40:42),'(I3)')NF2
        LCA=42
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 40 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          IF(ISWIT1.EQ.1) THEN
            YPL(I)=SDFTST(XPL(I),NF1,NF2)
          ELSE IF(ISWIT1.EQ.2) THEN
            YPL(I)=SCFTST(XPL(I),NF1,NF2)
          END IF
40      CONTINUE
      ELSE IF(ISWIT.EQ.5) THEN
C t distribution
        WRITE(*,*)' Enter n (>= 1)'
        WRITE(*,*)' >'
        READ(*,*)N
        XMIN=-3.D0
        XMAX=3.D0
        YMIN=0.D0
        IF(ISWIT1.EQ.1) THEN
          YMAX=0.5D0
        ELSE
          YMAX=1.D0
        END IF
        TX='t'
        LTX=1
        IF(ISWIT1.EQ.1) THEN
          TY='f(t)'
        ELSE IF(ISWIT1.EQ.2) THEN
          TY='F(t)'
        END IF
        LTY=4
        CAPT='t distribution with n = ***'
        WRITE(CAPT(25:27),'(I3)')N
        LCA=27
        DEL=(XMAX-XMIN)/DBLE(NPL-1)
        DO 50 I=1,NPL
          XPL(I)=XMIN+(I-1)*DEL
          IF(ISWIT1.EQ.1) THEN
            YPL(I)=SDSTUD(XPL(I),N)
          ELSE IF(ISWIT1.EQ.2) THEN
            YPL(I)=SCSTUD(XPL(I),N)
          END IF
50      CONTINUE
      END IF
C graphics
      CALL GROPEN
      CALL GROPWS(NWS)
      CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
      CALL GRVWWC(-0.15D0,0.9D0,0.15D0,0.85D0)
      CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
      CALL GRFRAM
      CALL GRBOUN
      CALL GRSCLX(TX,LTX)
      CALL GRSCLY(TY,LTY)
      CALL GRTXTC(1.D0,CAPT,LCA)
      CALL GRSTCL(2)
      CALL GRPLIN(NPL,XPL,YPL)
      CALL GRCLSE
      END

File: runme/mcgen/local/datan/solsrc/s2sd.c
#include <stdio.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__1000 = 1000;
static double c_dp999 = .999, c_dmp15 = -.15, c_dp9 = .9, c_dp15 = .15,
              c_dp85 = .85, c_dmp414 = -.414, c_d1 = 1., c_d0 = 0.;

main(void)
{
    /* Local variables */
    static integer i, n, iswit1, iswit, lca, ltx, lty, nws, nf1, nf2;
    static double del, xpl[1000], ypl[1000], sigma, x0, xmin, ymin, xmax, ymax;
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    double d__1;
/* identify program to user */
    printf("%s\n", "Program S2D provides interactively graphs of");
    printf("%s\n\n", "statistical functions of a continuous variable.");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - standardized normal distribution");
    printf("%s\n", "2 - normal distribution");
    printf("%s\n", "3 - chi-squared distribution");
    printf("%s\n", "4 - F distribution");
    printf("%s\n", "5 - t distribution");
    printf("%s", "> ");
    scanf("%li", &iswit);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - probability density");
    printf("%s\n", "2 - distribution function");
    printf("%s", "> ");
    scanf("%li", &iswit1);
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    if (iswit == 1) {
/* standardized normal */
        xmin = -3.;
        xmax = 3.;
        ymin = 0.;
        if (iswit1 == 1) {
            ymax = .5;
        } else {
            ymax = 1.;
        }
        strncpy(tx, "x", 1);
        ltx = 1;
        if (iswit1 == 1) {
            strncpy(ty, "&f@(x)", 6);
        } else if (iswit1 == 2) {
            strncpy(ty, "&F@(x)", 6);
        }
        lty = 6;
        strncpy(capt, "standardized normal distribution", 32);
        lca = 32;
        del = (xmax - xmin) / 999.;
        for (i = 1; i <= 1000; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            if (iswit1 == 1) {
                ypl[i - 1] = sdstnr_(&xpl[i - 1]);
            } else if (iswit1 == 2) {
                ypl[i - 1] = scstnr_(&xpl[i - 1]);
            }
        }
    } else if (iswit == 2) {
/* normal */
        printf("%s\n", "Enter x0:");
        printf("%s", "> ");
        scanf("%lf", &x0);
        printf("%s\n", "Enter sigma (> 0.):");
        printf("%s", "> ");
        scanf("%lf", &sigma);
        xmin = sigma * -3. + x0;
        xmax = sigma * 3. + x0;
        ymin = 0.;
        if (iswit1 == 1) {
            ymax = sdnorm_(&x0, &x0, &sigma) * 1.2;
        } else {
            ymax = 1.;
        }
        strncpy(tx, "x", 1);
        ltx = 1;
        if (iswit1 == 1) {
            strncpy(ty, "f(x)", 4);
        } else if (iswit1 == 2) {
            strncpy(ty, "F(x)", 4);
        }
        lty = 4;
        sprintf(capt, "normal distribution with x_0# = %6.2lf, &s@ = %6.2lf",
                x0, sigma);
        capt[52]= ' ';
        lca = 52;
        del = (xmax - xmin) / 999.;
        for (i = 1; i <= 1000; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            if (iswit1 == 1) {
                ypl[i - 1] = sdnorm_(&xpl[i - 1], &x0, &sigma);
            } else if (iswit1 == 2) {
                ypl[i - 1] = scnorm_(&xpl[i - 1], &x0, &sigma);
            }
        }
    } else if (iswit == 3) {
/* chi-squared */
        printf("%s\n", "Enter n (>= 1):");
        printf("%s", "> ");
        scanf("%li", &n);
        xmin = 0.;
        xmax = sqchi2_(&c_dp999, &n);
        if (n == 1) {
            xmin = 1e-5;
        }
        ymin = 0.;
        if (iswit1 == 1) {
            if (n == 1) {
                ymax = 2.;
            }
            if (n == 2) {
                ymax = .5;
            }
            if (n > 2) {
                d__1 = (double) (n - 1);
                ymax = sdchi2_(&d__1, &n) * 1.2;
            }
        } else if (iswit1 == 2) {
            ymax = 1.;
        }
        strncpy(tx, "&c^2", 4);
        ltx = 4;
        if (iswit1 == 1) {
            strncpy(ty, "f(&c@^2#)", 9);
        } else if (iswit1 == 2) {
            strncpy(ty, "F(&c@^2#)", 9);
        }
        lty = 9;
        sprintf(capt, "&c@^2# distribution with n = %3li", n);
        capt[32] = ' ';
        lca = 32;
        del = (xmax - xmin) / 999.;
        for (i = 1; i <= 1000; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            if (iswit1 == 1) {
                ypl[i - 1] = sdchi2_(&xpl[i - 1], &n);
            } else if (iswit1 == 2) {
                ypl[i - 1] = scchi2_(&xpl[i - 1], &n);
            }
        }
    } else if (iswit == 4) {
/* F distribution */
        printf("%s\n", "Enter f1 (>= 1):");
        printf("%s", "> ");
        scanf("%li", &nf1);
        printf("%s\n", "Enter f2 (>= 1):");
        printf("%s", "> ");
        scanf("%li", &nf2);
        xmin = 0.;
        xmax = 3.;
        ymin = 0.;
        ymax = 1.;
        strncpy(tx, "F", 1);
        ltx = 1;
        if (iswit1 == 1) {
            strncpy(ty, "f(F)", 4);
        } else if (iswit1 == 2) {
            strncpy(ty, "F(F)", 4);
        }
        lty = 4;
        sprintf(capt, "F distribution with f_1# = %3li, f_2# = %3li",
                nf1, nf2);
        capt[42] = ' ';
        lca = 42;
        del = (xmax - xmin) / 999.;
        for (i = 1; i <= 1000; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            if (iswit1 == 1) {
                ypl[i - 1] = sdftst_(&xpl[i - 1], &nf1, &nf2);
            } else if (iswit1 == 2) {
                ypl[i - 1] = scftst_(&xpl[i - 1], &nf1, &nf2);
            }
        }
    } else if (iswit == 5) {
/* t distribution */
        printf("%s\n", "Enter n (>=1):");
        printf("%s", "> ");
        scanf("%li", &n);
        xmin = -3.;
        xmax = 3.;
        ymin = 0.;
        if (iswit1 == 1) {
            ymax = .5;
        } else {
            ymax = 1.;
        }
        strncpy(tx, "t", 1);
        ltx = 1;
        if (iswit1 == 1) {
            strncpy(ty, "f(t)", 4);
        } else if (iswit1 == 2) {
            strncpy(ty, "F(t)", 4);
        }
        lty = 4;
        sprintf(capt, "t distribution with n = %3li", n);
        capt[27] = ' ';
        lca = 27;
        del = (xmax - xmin) / 999.;
        for (i = 1; i <= 1000; ++i) {
            xpl[i - 1] = xmin + (i - 1) * del;
            if (iswit1 == 1) {
                ypl[i - 1] = sdstud_(&xpl[i - 1], &n);
            } else if (iswit1 == 2) {
                ypl[i - 1] = scstud_(&xpl[i - 1], &n);
            }
        }
    }
    (void)getchar();
/* graphics */
    gropen_();
    gropws_(&nws);
    grwncc_(&xmin, &xmax, &ymin, &ymax);
    grvwwc_(&c_dmp15, &c_dp9, &c_dp15, &c_dp85);
    grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
    grfram_();
    grboun_();
    grsclx_(tx, &ltx);
    grscly_(ty, &lty);
    grtxtc_(&c_d1, capt, &lca);
    grstcl_(&c__2);
    grplin_(&c__1000, xpl, ypl);
    grclse_();
    return 0;
} /* main */

